platform/linux-generic/test/validation/api/ml/ml_linux.c

```diff
@@ -0,0 +1,1259 @@
+		global.disabled = 1;
+		ODPH_DBG("ML test disabled\n");
+		return 0;
+	}
+
+	/* Configure ML */
+	odp_ml_config_init(&global.ml_config);
+	global.ml_config.max_models_created = ml_capa->max_models;
+	global.ml_config.max_models_loaded = ml_capa->max_models_loaded;
+	global.ml_config.max_model_size = CONFIG_MAX_MODEL_SIZE;
+
+	if (ml_capa->load.compl_mode_mask & ODP_ML_COMPL_MODE_SYNC)
+		global.ml_config.load_mode_mask |= ODP_ML_COMPL_MODE_SYNC;
+
+	if (ml_capa->load.compl_mode_mask & ODP_ML_COMPL_MODE_POLL)
+		global.ml_config.load_mode_mask |= ODP_ML_COMPL_MODE_POLL;
+
+	if (ml_capa->load.compl_mode_mask & ODP_ML_COMPL_MODE_EVENT)
+		global.ml_config.load_mode_mask |= ODP_ML_COMPL_MODE_EVENT;
+
+	if (ml_capa->run.compl_mode_mask & ODP_ML_COMPL_MODE_SYNC)
+		global.ml_config.run_mode_mask |= ODP_ML_COMPL_MODE_SYNC;
+
+	if (ml_capa->run.compl_mode_mask & ODP_ML_COMPL_MODE_POLL)
+		global.ml_config.run_mode_mask |= ODP_ML_COMPL_MODE_POLL;
+
+	if (ml_capa->run.compl_mode_mask & ODP_ML_COMPL_MODE_EVENT)
+		global.ml_config.run_mode_mask |= ODP_ML_COMPL_MODE_EVENT;
+
+	if (odp_ml_config(&global.ml_config)) {
+		ODPH_ERR("\n\nConfiguring ML failed\n");
+		return -1;
+	}
+
+	global.x = 5;
+	global.wait_ns = 500 * ODP_TIME_MSEC_IN_NS;
+	global.y_expected = 19; /* y = 3 * x + 4 = 3 * 5 + 4 = 19 */
+
+	/* Prepare data for running model inference */
+	odp_ml_run_param_init(&global.run_param);
+
+	global.data.num_input_seg = NUM_INPUTS;
+	global.data.input_seg = &global.input_seg;
+	global.input_seg.size = sizeof(int32_t);
+	global.input_seg.addr = &global.x;
+
+	global.data.num_output_seg = NUM_OUTPUTS;
+	global.data.output_seg = &global.output_seg;
+	global.output_seg.size = sizeof(int32_t);
+	global.output_seg.addr = &global.y;
+
+	if (fill_model_param("simple_linear.onnx", &global.model_param))
+		return -1;
+
+	/* Create ML model */
+	global.ml_model = odp_ml_model_create(MODEL_NAME, &global.model_param);
+	if (global.ml_model == ODP_ML_MODEL_INVALID) {
+		ODPH_ERR("Create ML model failed\n");
+		free(global.model_param.model);
+		return -1;
+	}
+
+	/* Asynchronous mode with event completion is not supported */
+	if (!((ml_capa->load.compl_mode_mask & ODP_ML_COMPL_MODE_EVENT) ||
+	      (ml_capa->run.compl_mode_mask & ODP_ML_COMPL_MODE_EVENT)))
+		return 0;
+
+	/* Create a queue for sending ML completion event to */
+	odp_queue_param_init(&queue_param);
+	queue_param.type        = ODP_QUEUE_TYPE_SCHED;
+	queue_param.sched.sync  = ODP_SCHED_SYNC_PARALLEL;
+	queue_param.sched.prio  = odp_schedule_default_prio();
+	queue_param.sched.group = ODP_SCHED_GROUP_ALL;
+
+	global.queue = odp_queue_create("ML compl queue", &queue_param);
+	if (global.queue == ODP_QUEUE_INVALID) {
+		ODPH_ERR("Queue create failed\n");
+		free(global.model_param.model);
+		return -1;
+	}
+
+	/* Create an ML job completion pool */
+	if (ml_capa->pool.max_num < NUM_COMPL) {
+		ODPH_ERR("Too small ML compl pool %u\n", ml_capa->pool.max_num);
+		return -1;
```

Comment:
Error branch missing memory free "free(global.model_param.model);". Maybe simpler to use "goto error" in all branches, and have the memory free there once.

Reply:
Did the "goto error" thing in v3.